docs/source/configurations.md
# Configurations

This document provides an overview of the `config.yaml` file used for configuring the project and explains the available command-line options for overriding configurations.

## Overview of `config.yaml`

The `config.yaml` file organizes the project configuration into three main sections: `DATA`, `directories`, and `training`.

```{note}
The YAML file is **not required** for running the project with default settings. The project is designed to run seamlessly without a `config.yaml` file by using pre-defined default configuration values.
```

---

## Default configurations

### Data Configuration

The `DATA` section specifies the dataset and its properties:

```yaml
dataset:
  dataset_path: "erikpinhasov/landcover_dataset"
  batch_size: 16
  num_workers: 8
  do_reduce_labels: False
  pin_memory: True
```

**Details:**

- `dataset_path`: Path to the dataset. In this case, it references a Hugging Face dataset.
- `batch_size`: Number of samples processed per training batch.
- `num_workers`: Number of worker threads for data loading.
- `do_reduce_labels`: Whether to apply label reduction.
- `pin_memory`: If `True`, enables faster GPU transfer by pinning memory.
- `id2label`: Maps class indices to their corresponding labels.

---

### Project Configuration

The `directories` section defines directory paths for saving models, logs, and results:

```yaml
directories:
  models: models
  pretrained: models/pretrained_models
  logs: models/logs
  checkpoints: models/logs/checkpoints
  results: results
```

**Details:**

- `models`: Directory to save trained models.
- `pretrained`: Directory for storing pre-trained models.
- `logs`: Directory for logs.
- `checkpoints`: Directory for storing training checkpoints.
- `results`: Directory for evaluation results.

---

### Training Configuration

The `training` section provides options for training parameters:

```yaml
training:
  model_name: b0
  max_epochs: 50
  learning_rate: 2e-5
  weight_decay: 1e-3
  ignore_index: 0
  weighting_strategy: "raw"
  alpha: 0.7
  id2label:
    0: "background"
    1: "bareland"
    2: "rangeland"
    3: "developed space"
    4: "road"
    5: "tree"
    6: "water"
    7: "agriculture land"
    8: "buildings"
```

**Details:**

- `model_name`: Name of the SegFormer variant (e.g., `b0`).
- `max_epochs`: Maximum number of epochs for training.
- `learning_rate`: Learning rate for the optimizer.
- `weight_decay`: Strength of L2 regularization applied to the optimizer.
- `ignore_index`: Label index to ignore during training (`int` or `None`).
- `weighting_strategy`: Method for normalizing class weights (`"none"`, `"balanced"`, `"max"`, `"sum"`, `"raw"`).
- `alpha`: Cross-Entroy loss alpha weight parameter, beta = 1 - alpha.
- `id2label`: Mapping from class indices to class labels.

---

### Callbacks Configuration

The `callbacks` section defines parameters for managing callbacks during training:

```yaml
callbacks:
  early_stop_patience: 5
  early_stop_monitor: "val_loss"
  early_stop_mode: "min"
  save_model_monitor: "val_mean_iou"
  save_model_mode: "max"
```

**Details:**

- `early_stop_patience`: Number of epochs to wait for improvement before stopping training.
- `early_stop_monitor`: Metric to monitor for early stopping (e.g., `"val_loss"`).
- `early_stop_mode`: Direction of improvement (`"min"` for decreasing metrics, `"max"` for increasing metrics).
- `save_model_monitor`: Metric to track for saving the best model (e.g., `"val_mean_iou"`).
- `save_model_mode`: Direction of improvement (`"min"` for decreasing metrics, `"max"` for increasing metrics).

---

## Command-Line Options

Command-line options allow overriding `config.yaml` settings or default values. For example:

- Specify a custom configuration file:

  ```bash
  ucs --config path/to/config.yaml train
  ```

- Override specific parameters without a configuration file:
  ```bash
  ucs train --batch_size 32 --learning_rate 0.001
  ```

---

For more details, refer to the [Usage Guide](./usage.md).


_sources/configurations.rst.txt
Configurations
===============

src/ucs/utils/config.py
from dataclasses import field
from pathlib import Path
from typing import Dict, Optional

import yaml
from pydantic.dataclasses import dataclass


@dataclass
class CallbacksConfig:
    early_stop_patience: int = 5
    early_stop_monitor: str = "val_loss"
    early_stop_mode: str = "min"
    save_model_monitor: str = "val_mean_iou"
    save_model_mode: str = "max"


@dataclass
class TrainingConfig:  # pylint: disable=too-many-instance-attributes
    model_name: str = "b0"
    max_epochs: int = 50
    learning_rate: float = 2e-5
    weight_decay: float = 1e-3
    ignore_index: Optional[int] = 0  # int | None
    # Options: "none", "balanced", "max", "sum", or "raw"
    weighting_strategy: str = "raw"
    gamma: float = 2
    id2label: Dict[int, str] = field(
        default_factory=lambda: {
            0: "background",
            1: "bareland",
            2: "rangeland",
            3: "developed space",
            4: "road",
            5: "tree",
            6: "water",
            7: "agriculture land",
            8: "buildings",
        }
    )


@dataclass
class DirectoriesConfig:
    models: str = "models"
    pretrained: str = "models/pretrained_models"
    logs: str = "models/logs"
    checkpoints: str = "models/logs/checkpoints"
    results: str = "results"


@dataclass
class DatasetConfig:
    dataset_path: str = "erikpinhasov/landcover_dataset"
    batch_size: int = 16
    num_workers: int = 8
    do_reduce_labels: bool = False
    pin_memory: bool = True
    model_name: Optional[str] = None


@dataclass
class Config:
    dataset: DatasetConfig = field(default_factory=DatasetConfig)
    directories: DirectoriesConfig = field(default_factory=DirectoriesConfig)
    training: TrainingConfig = field(default_factory=TrainingConfig)
    callbacks: CallbacksConfig = field(default_factory=CallbacksConfig)

    @classmethod
    def load_config(cls, config_path: Optional[str] = None, **overrides) -> "Config":
        """
        Load YAML configuration file and apply overrides.

        Args:
            config_path (Optional[str]): Path to the configuration YAML file.
            **overrides: Dictionary of command-line overrides.

        Returns:
            Config: An instance of the Config class with loaded values.
        """
        config = cls()

        if config_path and Path(config_path).resolve().exists():
            with open(config_path, "r", encoding="utf8") as file:
                yaml_data = yaml.safe_load(file) or {}
            config = cls(**yaml_data)

        config.dataset.model_name = config.training.model_name
        config._apply_overrides(overrides)
        config._create_directories()
        return config

    def _apply_overrides(self, overrides):
        """
        Applies CLI override values to the configuration.

        Args:
            overrides (dict): Dictionary of values to override.
        """

        for key, value in overrides.items():
            if value is None:
                continue

            for section_name in self.__annotations__:  # Iterate over attributes
                section = getattr(self, section_name)

                if hasattr(
                    section, key
                ):  # Check if the override key exists in the section
                    setattr(section, key, value)
                    break

    def _create_directories(self):
        """
        Creates required directories if they do not exist.

        Prints the directories that were created.
        """

        created_dirs = [
            dir_path
            for dir_path in vars(self.directories).values()
            if not Path(dir_path).exists()
            and Path(dir_path).mkdir(parents=True, exist_ok=True) is None
        ]
        if created_dirs:
            print("Created directories:\n" + "\n".join(created_dirs))
